pkg/api/validate/openshiftcluster_validatedynamic.go

```diff
@@ -150,13 +156,67 @@ func (dv *openShiftClusterDynamicValidator) validateServicePrincipalProfile(ctx
 	return autorest.NewBearerAuthorizer(token), nil
 }
 
-func (dv *openShiftClusterDynamicValidator) validateVnetPermissions(ctx context.Context, oc *api.OpenShiftCluster, client authorization.PermissionsClient, code, typ string) error {
-	vnetID, _, err := subnet.Split(oc.Properties.MasterProfile.SubnetID)
+func (dv *openShiftClusterDynamicValidator) validateVnetPermissions(ctx context.Context, vnet mgmtnetwork.VirtualNetwork, oc *api.OpenShiftCluster, client authorization.PermissionsClient, code, typ string) error {
+	r, err := azure.ParseResourceID(*vnet.ID)
 	if err != nil {
 		return err
 	}
 
-	r, err := azure.ParseResourceID(vnetID)
+	timeoutCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+
+	err = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
+		permissions, err := client.ListForResource(ctx, r.ResourceGroup, r.Provider, r.ResourceType, "", r.ResourceName)
+		if detailedErr, ok := err.(autorest.DetailedError); ok &&
+			detailedErr.StatusCode == http.StatusForbidden {
+			return false, nil
+		}
+		if err != nil {
+			return false, err
+		}
+
+		for _, action := range []string{
+			"Microsoft.Network/virtualNetworks/subnets/join/action",
+			"Microsoft.Network/virtualNetworks/subnets/read",
+			"Microsoft.Network/virtualNetworks/subnets/write",
+		} {
+			ok, err := utilpermissions.CanDoAction(permissions, action)
+			if err != nil {
+				return false, err
+			}
+			if !ok {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}, timeoutCtx.Done())
+	if err == wait.ErrWaitTimeout {
+		return api.NewCloudError(http.StatusBadRequest, code, "", "The "+typ+" does not have Contributor permission on vnet '%s'.", *vnet.ID)
+	}
+	if detailedErr, ok := err.(autorest.DetailedError); ok &&
+		detailedErr.StatusCode == http.StatusNotFound {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidLinkedVNet, "", "The vnet '%s' could not be found.", *vnet.ID)
+	}
+	if err != nil {
+		return err
+	}
+
+	// validate route table permissions
+	for _, sn := range *vnet.VirtualNetworkPropertiesFormat.Subnets {
+		if sn.RouteTable != nil {
```

Comment:
nit, you can say `if sn.RouteTable == nil { continue }` and save yourself an indent